File: Practice/37/Python/Zadanie_37.py
from math import pi as pi;from math import cos as cos;from math import sin  as sin
from math import sqrt as sqrt;from math import atan2 as atan
from math import hypot as hypot
import enum
class Cords(enum.IntEnum):
    Cartesian = 1
    Polar = enum.auto()
class Point():

    def __init__(self, a1=0, a2=0, coord_system=Cords.Cartesian):

        self.R, self.EPS = None, 10**-10 # self.R=None индикатор, что его надо рассчитать

        if isinstance(a1, str):
            try: self.x,  self.y = map(float, a1[1:-1].split(','))
            except: #raise TypeError("Only integers are allowed")
                self.x = self.y = 0
                print("Неверный формат координат. Взяты предустановленные значения.")
        elif isinstance(a1, (int, float)) and isinstance(a2, (int, float)):
            if coord_system==Cords.Cartesian: # not 'Polar'
                self.x,  self.y = a1, a2
            else:
                self.x, self.y, self.R, self.ang = a1*cos(a2), a1*sin(a2), a1, a2
        else:
            self.x = self.y = 0
            print("Неверный формат координат. Взяты предустановленные значения.")

        if self.R == None: self.CartesianToPolar()

    def PolarToCartesian(self):
        self.x, self.y = self.R*cos(self.ang), self.R*sin(self.ang)

    def CartesianToPolar(self):
        self.R   = sqrt(self.x**2 + self.y**2)
        self.ang = atan(self.y, self.x)

    def __eq__(self, Other): #==
        return abs(self.x - Other.x)<=self.EPS and abs(self.y - Other.y)<=self.EPS

    def __ne__(self, Other): #!=
        return abs(self.x - Other.x)>self.EPS or abs(self.y - Other.y)>self.EPS

    def __str__(self):
        return "(" + str(self.x) + "," + str(self.y) + ")"

    def __repr__(self):
        return "(" + str(self.x) + "," + str(self.y) + ")"

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def get_r(self):
        return self.R

    def get_phi(self):
        return self.ang

    def set_x(self, x):
        self.x = x
        self.CartesianToPolar()

    def set_y(self, y):
        self.y = y
        self.CartesianToPolar()

    def set_r(self, R):
        self.R = R
        self.PolarToCartesian()

    def set_phi(self, ang):
        self.ang = ang
        self.PolarToCartesian()

#a = Vector(Point(1, 2))
#b = Vector(Point(-2, 0), Point(-1, 2))
#Если аргументы не заданы, то создаётся вектор единичной длины в направлении оси
#Если задан один аргумент, создаётся вектор направленный из начала координат в точку переданную в качестве параметра.
#Если заданы оба аргументаbegin и end, то создаётся вектор направленный из begin в end.
class Vector():
    def __init__(self, a=None, b=None ):
        if not isinstance(a, Point) and not isinstance(b, Point):
            self.begin = Point(0.0, 0.0)
            self.end = Point(1.0, 0.0)
        elif not isinstance(b, Point):
            self.begin = Point(0.0, 0.0)
            self.end = a
        else:
            self.begin = a
            self.end = b

        self.cord = [self.end.x - self.begin.x, self.end.y - self.begin.y]  # vector coordinates
        self.len  = hypot(*self.cord)                                       # fast len calculation
        self.ang  = atan(self.cord[1], self.cord[0])                        # angel in radians

    def __repr__(self):
        return "[" + str(self.begin) + "," + str(self.end) + "]=(" + str(self.cord[0]) + "," + str(self.cord[1]) + ")"

    def __str__(self):
        return "[" + str(self.begin) + "," + str(self.end) + "]=(" + str(self.cord[0]) + "," + str(self.cord[1]) + ")"

    def __eq__(self, Other): #== #__eq__ (оператор ==). Сравнивает два объекта класса Vector на равенство. Точность сравнения 10-10.
        return self.cord == Other.cord

    def length(self): #length. Метод не принимает параметров и возвращает длину вектора;
        return self.len

    def __mul__(self, Other):
        if isinstance(Other, (int, float)):             #__mul__ (бинарный *). Левый операнд класса Vector, правый типа double.  Возвращает новый вектор компоненты которого умножены на правый операнд.
            return Vector(a=self.begin, b=Point(self.begin.x+self.cord[0]*Other, self.begin.y+self.cord[1]*Other))
        elif isinstance(Other, Vector):                 #__mul__ (бинарный *). Левый и правый операнды класса Vector.  Возвращает вещественное число вычисляющееся как скалярное произведение векторов-операндов.
            return self.len * Other.len * cos(self.ang-Other.ang)
        return self                             # просто возвращаем свой вектор без изменений: ошибки не идет.

    def __neg__(self): #__neg__ (унарный -). Возвращает новый вектор направление которого противоположно направлению вектора к которому применяется оператор.
        return self.__mul__(-1)
        #Vector(begin=self.begin, end=Point(self.begin.x-self.cord[0], self.begin.y-self.cord[1]))

    #__add__ (бинарный +). Левый и правый операнды класса Vector.  новый вектор являющийся покомпонентной суммой левого и правого операндов.
    def __add__(self, Other): # в условии не указано, чью точку begin брать за начало нового вектора. Берем self.begin
        return Vector(a=self.begin, b=Point(self.end.x+Other.cord[0], self.end.y+Other.cord[1]))

    #__sub__ (бинарный -). Левый и правый операнды класса Vector.  новый вектор являющийся покомпонентной разностью левого и правого операндов.
    def __sub__(self, Other):
        return Vector(a=self.begin, b=Point(self.end.x-Other.cord[0], self.end.y-Other.cord[1]))
        # return self.__add__(Other.__neg__())


### Test 37 part ######################################################
def test_37():
    import math
    def equal(a, b, e=1E-10):
        if -e < a - b < e: return True
        else: return False

    a = Vector(Point(1, 2))
    b = Vector(Point(-2, 0), Point(-1, 2))

    if a == b and b == a: print('Equality test passed')
    else: print('Equality test failed')

    na  = Vector(Point(-1, -2))
    ox  = Vector(Point( 1,  0))
    nox = Vector(Point(-1,  0))
    oy  = Vector(Point( 0,  1))
    noy = Vector(Point( 0, -1))

    if a == -na and na == -a and -ox == nox and -oy == noy: print('Invert test passed')
    else: print('Invert test failed')

    if ox + oy + oy == a and -ox == -a + oy + oy: print('Summation test passed')
    else: print('Summation test failed')

    if -ox + oy == oy - ox and -oy + ox == ox - oy: print('Subtraction test passed')
    else: print('Subtraction test failed')

    if (ox * 3 == ox + ox + ox and
        oy * 3 == oy + oy + oy and
        ox * (-3) == -ox - ox - ox and
        oy * (-3) == -oy - oy - oy): print('Multiplication by number test passed')
    else: print('Multiplication by number test failed')

    if (equal(ox.length(), 1) and
        equal(oy.length(), 1) and
        equal((ox * 3 + oy * 4).length(), 5)): print('Length test passed')
    else: print('Length test failed')

    if (equal(ox*ox, ox.length()**2) and
        equal(oy*oy, oy.length()**2) and
        equal((ox*3 + oy*4)*(ox*3 + oy*4), (ox*3 + oy*4).length()**2)): print('Multiplication by Vector test passed')
    else: print('Multiplication by Vector test failed')

### Test 36 part ######################################################
def test_36():
    import copy
    with open('data.txt') as fin:
        original = [Point(p) for p in fin.readline().split(', ')]
    simulacrum = copy.deepcopy(original)

    for p in simulacrum:
        print(p, end='')
        p.set_x(p.get_x() + 10)
        p.set_phi(p.get_phi() + 180*pi/180)
        p.set_y(-p.get_y())
        p.set_x(-p.get_x() - 10)
        print(p)

    print('\nIt works!\n' if simulacrum == original else '\nIt not works!\n')

test_37()

File: Practice/37/C++/Zadanie_37.cpp
#include <iostream>
#include <cmath> 
#include <string>
#include <fstream>
#include <vector>

using namespace std;


double sqr(double a);
bool equal(double a, double b, double e = 1E-10);

const double PI = 3.141592653589793;

enum Cords { Cartesian, Polar };

class Point {

private:
	static const double EPS;	// возможное отклонение
	double x, y, r, ang;

	void CartesianToPolar() {
		this->r = hypot(this->x, this->y);
		this->ang = atan2(this->y, this->x);
	}

	void PolarToCartesian() {
		this->x = this->r * cos(this->ang);
		this->y = this->r * sin(this->ang);
	}

public:
	Point(double a1 = 0, double a2 = 0, Cords coord_system = Cartesian) {
		if (coord_system == Cartesian) {	// Cartesian
			this->x = a1;
			set_y(a2);
		}
		else {								// Polar
			this->r = a1;
			set_phi(a2);
		}
	}

	bool operator == (Point& Other) { // const 
		return (abs(this->x - Other.get_x()) < EPS and
			abs(this->y - Other.get_y()) < EPS);
	}

	bool operator != (Point& Other) { // const 
		return not (abs(this->x - Other.get_x()) < EPS and
			abs(this->y - Other.get_y()) < EPS);
	}

	double get_x()   const { return x; }

	double get_y()   const { return y; }

	double get_r()   const { return r; }

	double get_phi() const { return ang; }

	void set_x(double x_) {
		this->x = x_;
		CartesianToPolar();
	}

	void set_y(double y_) {
		this->y = y_;
		CartesianToPolar();
	}

	void set_r(double r_) {
		this->r = r_;
		PolarToCartesian();
	}

	void set_phi(double phi) {
		this->ang = phi;
		PolarToCartesian();
	}
};
const double Point::EPS = pow(10, -10);

std::ostream& operator<< (ostream& out, Point& point) { //const 
	return out << "(" << point.get_x() << "," << point.get_y() << ")";
}

std::istream& operator>> (istream& istrm, Point& point) {
	string inp;
	istrm >> inp;
	std::size_t f1 = inp.find('(') + 1;
	std::size_t f2 = inp.find(',');
	std::size_t f3 = inp.find(')') - 1;

	if ((f1 == string::npos) || (f2 == string::npos) || (f3 == string::npos)) {
		return istrm; // ошибка формата строки ввода
	}

	point.set_x(stod(inp.substr(f1, f2 - f1)));
	point.set_y(stod(inp.substr(f2 + 1, f3 - f2)));

	return istrm;
}

// Задание 37 //////////////////////////////////////////////////////////////////

class Vector {

private:
	const double EPS = pow(10, -10);;	// возможное отклонение
	Point begin, end;
	void cords() {	this->cordX = this->end.get_x() - this->begin.get_x();
					this->cordY = this->end.get_y() - this->begin.get_y();}

public:
	Vector() :						begin(Point(0, 0)), end(Point(1, 0)) {this->cords();}
	Vector(Point end):				begin(Point(0, 0)), end(end)		 {this->cords();}
	Vector(Point begin, Point end): begin(begin),		end(end)		 {this->cords();}
	double cordX, cordY;

	bool operator == (const Vector& Other) {
		return (abs(this->cordX - Other.cordX) < EPS) &&
			   (abs(this->cordY - Other.cordY) < EPS);
	}
	Vector operator- () {
		Vector vect(this->begin, Point(this->begin.get_x()-cordX, this->begin.get_y() - cordY) );
		return vect;
	}
	Vector operator+ (const Vector& Other) {
		Point vect_end(this->end.get_x() + Other.cordX, this->end.get_y() + Other.cordY);
		Vector vect(this->begin, vect_end);
		return vect;
	}
	//Vector operator- (const Vector& Other) {
	//	return this + Other;
	//}
	
	Vector operator- (const Vector& Other) {
		Point vect_end(this->end.get_x() - Other.cordX, this->end.get_y() - Other.cordY);
		Vector vect(this->begin, vect_end);
		return vect;
	}

	Vector operator* (double Other) {
		Point vect_end(this->begin.get_x() + this->cordX * Other,
					   this->begin.get_y() + this->cordY * Other);
		Vector vect(this->begin, vect_end);
		return vect;
	}
	double length() { return hypot(this->cordX, this->cordY); }
	
	//double operator* (const Vector& Other) {
	//	double Other_len = Other.   .length();
	//	double this_ang = atan2(this->cordY, this->cordX);
	//	double Other_ang= atan2(Other.cordY, Other.cordX);
	//	return this->length() * Other.length() * cos(this_ang- Other_ang);
	//}
	
	double operator* (const Vector& Other) {
		double Other_len = hypot(Other.end.get_x() - Other.begin.get_x(), Other.end.get_y() - Other.begin.get_y());
		double scalar = (this->length()) * Other_len * cos(atan2((this->end.get_y() - this->begin.get_y()), (this->end.get_x() - this->begin.get_x())) - atan2((Other.end.get_y() - Other.begin.get_y()), (Other.end.get_x() - Other.begin.get_x())));
		return scalar;
	}
};

int main()
{
	Vector a(Point(1, 2)), b(Point(-2, 0), Point(-1, 2));

	if (a == b ) cout << "Equality test passed\n"; //&& b == a
	else cout << "Equality test failed\n";

	Vector na(Point(-1, -2)), ox(Point(1, 0)), nox(Point(-1, 0)), oy(Point(0, 1)), noy(Point(0, -1));

	if (a == -na && na == -a && -ox == nox && -oy == noy) cout << "Invert test passed\n";
	else cout << "Invert test failed\n";

	if (ox + oy + oy == a && -ox == -a + oy + oy) cout << "Summation test passed\n";
	else cout << "Summation test failed\n";

	if (-ox + oy == oy - ox && -oy + ox == ox - oy) cout << "Subtraction test passed\n";
	else cout << "Subtraction test failed\n";
	
	if (ox * 3 == ox + ox + ox &&
		oy * 3 == oy + oy + oy &&
		ox * (-3) == -ox - ox - ox &&
		oy * (-3) == -oy - oy - oy) cout << "Multiplication by number test passed\n";
	else cout << "Multiplication by number test failed\n";
	
	if (equal(ox.length(), 1) &&
		equal(oy.length(), 1) &&
		equal((ox * 3 + oy * 4).length(), 5)) cout << "Length test passed\n";
	else cout << "Length test failed\n";

	if (equal(ox * ox, sqr(ox.length())) &&
		equal(oy * oy, sqr(oy.length())) &&
		equal((ox * 3 + oy * 4) * (ox * 3 + oy * 4), sqr((ox * 3 + oy * 4).length()))) cout << "Multiplication by Vector test passed\n";
	else cout << "Multiplication by Vector test failed\n";
}



bool equal(double a, double b, double e) {
	if (-e < a - b && a - b < e) return true;
	else return false;
}

double sqr(double a) {
	return a * a;
}